### PR TITLE
Fix the proposal execution.

### PR DESCRIPTION
**Pull request by becketqin**
This patch changes the way executor carries out the proposals. Instead of doing on replica at a time, it take the entire partition as a single execution unit.

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskManager.java [hunk 1]
```diff
@@ -266,14 +267,13 @@ public void markTasksInProgress(List<ExecutionTask> tasks) {
         // Add task to the relevant task in progress.
         markTaskState(task, ExecutionTask.State.IN_PROGRESS);
         _inProgressPartitions.add(task.proposal.topicPartition());
-        if (task.proposal.balancingAction() == ActionType.REPLICA_MOVEMENT) {
-          if (task.sourceBrokerId() != null) {
-            _inProgressPartMovementsByBrokerId.put(task.sourceBrokerId(),
-                                                   _inProgressPartMovementsByBrokerId.get(task.sourceBrokerId()) + 1);
+        if (task.proposal.actionType() == ActionType.REPLICA_MOVEMENT) {
+          if (task.oldLeader() >= 0) {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskManager.java]: Should we clarify why we need this check?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskManager.java [hunk 2]
```diff
@@ -364,14 +364,13 @@ private void markTaskState(ExecutionTask task, ExecutionTask.State targetState)
    * Mark a given tasks as completed.
    */
   private void completeTask(ExecutionTask task) {
-    if (task.proposal.balancingAction() == ActionType.REPLICA_MOVEMENT) {
-      if (task.sourceBrokerId() != null) {
-        _inProgressPartMovementsByBrokerId.put(task.sourceBrokerId(),
-                                               _inProgressPartMovementsByBrokerId.get(task.sourceBrokerId()) - 1);
+    if (task.proposal.actionType() == ActionType.REPLICA_MOVEMENT) {
+      if (task.oldLeader() >= 0) {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskManager.java]: Should we clarify why we need this check?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskPlanner.java [hunk 3]
```diff
@@ -51,49 +49,55 @@ public ExecutionTaskPlanner() {
...
    *
+   * A proposal will have at least one of the two following actions:
+   * 1. Replica action (i.e. movement, addition, deletion or order change).
+   * 2. Leader action (i.e. leader movement)
+   *
    * @param proposal the proposal to execute.
    */
-  public void addBalancingProposal(BalancingAction proposal) {
+  public void addExecutionProposal(ExecutionProposal proposal) {
+    // Ideally we should avoid adjust replica order if not needed, but due to a bug in open source Kafka
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskPlanner.java]: Is there an open ticket that we can point to?

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/executor/ExecutorTest.java [hunk 4]
```diff
@@ -153,22 +167,22 @@ public void testBrokerDiesWhenMovePartitions() throws Exception {
...
-    for (BalancingAction proposal : proposalsToCheck) {
+    for (ExecutionProposal proposal : proposalsToCheck) {
       int replicationFactor = zkUtils.getReplicasForPartition(proposal.topic(), proposal.partitionId()).size();
       replicationFactors.put(new TopicPartition(proposal.topic(), proposal.partitionId()), replicationFactor);
     }
 
     long now = System.currentTimeMillis();
     while (executor.state().state() != ExecutorState.State.NO_TASK_IN_PROGRESS
-        && System.currentTimeMillis() < now + 5000) {
+        && System.currentTimeMillis() < now + 30000) {
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/executor/ExecutorTest.java]: Does 5000 fail?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskPlanner.java (hunk 3 above)
Comment (becketqin) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskPlanner.java]: It is fixed in latest Kafka.

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/executor/ExecutorTest.java (hunk 4 above)
Comment (becketqin) [reply cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/executor/ExecutorTest.java]: Yes...

**Pull request merged**